File: ipc/signals/python/receiver.py
import signal
import os
import time


def quit_handle(signum, frame):
    signame = signal.Signals(signum).name
    print('Finalizando...')
    exit(1)


def local_message(signum, frame):
    signame = signal.Signals(signum).name
    print(f'Recebendo sinal local {signame} [({signum}]')


def remote_message(signum, frame):
    signame = signal.Signals(signum).name
    print(f'Recebendo sinal remoto {signame} [{signum}]')


# registrando os tratadores dos sinais
signal.signal(signal.SIGINT, quit_handle)
signal.signal(signal.SIGALRM, local_message)
signal.signal(signal.SIGUSR1, remote_message)


print(f'Meu PID: {os.getpid()}')


while 1:
    signal.alarm(5)
    time.sleep(5)

File: ipc/signals/receiver.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <unistd.h>

void quit_handle(int s)
{
    printf("Finalizando...\n");
    exit(0);
}

void local_message(int s) 
{
     printf("Imprimindo mensagem local...\n" );
}

void remote_message(int s) 
{
     printf("Imprimindo mensagem remota...\n", s);
}

int main(void) {
    
    // registrando um tratador para o sinal
    signal(SIGINT, quit_handle);

    // registrando um tratador para o sinal
    signal(SIGALRM, local_message);

    // registrando um tratador para o sinal
    signal(SIGUSR1, remote_message);

    printf("Processo Receiver tem PID: %d\n",getpid());

    // faz um loop para manter o main ativo
    while (1)
    {
        // envia o SIGALRM a cada 5 segundos
        alarm(5);
        sleep(5);
    }

    return 0;
}
